classes/DirtEventTypes.sc

```diff
@@ -38,86 +38,93 @@ DirtEventTypes {
 	midiEvent = (
 			play: #{
 
-				var freq, lag, sustain, func;
-				var args, midiout, hasGate, midicmd, latency;
-				var sendNRPN;
+				var freq, lag, sustain;
+				var args, midiout, hasGate, midicmd, latency, chan;
+				var sendNRPN, schedmidi, schedmidicmd;
+				var hasNote = ~n != \none;
 
-				midicmd = ~midicmd;
 				midiout = ~midiout.value;
 
-				if(midicmd.isNil) {
-					if(~ccn.notNil) { midicmd = \control; ~ctlNum = ~ccn };
-					if(~ccv.notNil) { midicmd = \control; ~control = ~ccv };
-					if(~nrpn.notNil) {
-						var chan, nrpnMSB, nrpnLSB, valMSB, valLSB;
-						~val = ~val ? 0;
-						chan = ~midichan ? 0;
-						midicmd = \control;
-						nrpnLSB = ~nrpn % 128;
-						nrpnMSB = (~nrpn - nrpnLSB) / 128;
-						valLSB  = ~val % 128;
-						valMSB  = (~val - valLSB) / 128;
-						sendNRPN = {
-							midiout.control(chan, 99, nrpnMSB);
-							midiout.control(chan, 98, nrpnLSB);
-							midiout.control(chan, 6,  valMSB);
-							midiout.control(chan, 38, valLSB)
-						};
-					};
-					if(~progNum.notNil) { midicmd = \program };
-					if(~polyTouch.notNil) { midicmd = \polyTouch };
-					if(~midibend.notNil) { midicmd = \bend; ~val = ~midibend; };
-					if(~miditouch.notNil) { midicmd = \touch; ~val = ~miditouch; };
-					if(midicmd.isNil) { midicmd = \noteOn }; // if still nil
+				if(midiout.isNil) {
+					"midi device is nil.".postln;
+					return(true)
```

Comment:
hm? you probably mean 
```
^true
```

Reply:
I didn't want superdirt to keep looking for other event types. At this point I think right event type is found, there's just a problem with it. I haven't actually tested this control flow path, though..

Reply:
Tested now, and it seems `return` isn't a thing in sclang. I'll make a big if, as before.

Reply:
I think `^` is just SC syntax for `return`, right?

Reply:
Aha I thought it was `not`, fixed back to use `^true`

Reply:
I'm sorry, I didn't read carefully enough.

In methods of objects, you can return in the middle, with the caret `^`. But the functions here are not called as methods, but are functions in an event. So indeed, you can't use `^` here.

There are two ways to deal with this in these cases:

1. a "big if"
2. the method `block` allows you to jump out by a continuation-like construct. It passes a function to your outer function as first argument that you can use as an escape door by calling it with the return value:

```supercollider

(
x = block { |return|
	
	10.do { |i|
		if(0.4.coin) { return.value(i) };
	};
	
};
x.postln;
)
```
Maybe version (1) is easier in the present case.

Reply:
Hm, but if I change the code to this:
```
		if(midiout.isNil) {
					"midi device is nil.".postln;
					^true
				};
				"oh".postln;
```
Then it seems to work fine. At least, if I run `~dirt.soundLibrary.addMIDI(\midi2, nil);`, then I get `midi device is nil.`, but no `oh`. (The `oh` does appear as expected, when the midi device is not nil)

Reply:
Interesting. I think it just seems fine, but isn't. I'm pretty sure that you are exiting into the wrong context.